activation/activation.go

```diff
@@ -383,60 +324,57 @@ func (b *Builder) loadChallenge() error {
 	return nil
 }
 
-func (b *Builder) discardChallenge() error {
-	return b.store.Put(b.getNipstKey(), []byte{})
-}
-
-// PublishActivationTx attempts to publish an atx for the given epoch, it returns an error if an atx cannot be created
-// publish atx may not produce an atx each time it is called, that is expected behaviour as well.
-func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
-	if b.nipst != nil {
-		b.log.With().Info("re-entering atx creation in epoch %v", log.EpochId(uint64(epoch)))
+// PublishActivationTx attempts to publish an atx, it returns an error if an atx cannot be created.
+func (b *Builder) PublishActivationTx() error {
+	b.refreshChallenge()
+	if b.challenge != nil {
+		b.log.With().Info("using existing challenge", log.EpochId(uint64(b.currentEpoch())))
 	} else {
-		b.log.With().Info("starting build atx in epoch %v", log.EpochId(uint64(epoch)))
-		if b.challenge != nil {
-			if b.challenge.PubLayerIdx.GetEpoch(b.layersPerEpoch) < epoch {
-				log.Info("previous challenge loaded from store, but epoch has already passed, %s, %s", epoch, b.challenge.PubLayerIdx.GetEpoch(b.layersPerEpoch))
-				b.challenge = nil
-				err := b.discardChallenge()
-				if err != nil {
-					log.Error("cannot discard challenge")
-				}
-			}
-		}
-		if b.challenge == nil {
-			err := b.buildNipstChallenge(epoch)
-			if err != nil {
-				if _, alreadyPublished := err.(alreadyPublishedErr); alreadyPublished {
-					events.Publish(events.AtxCreated{false, "", uint64(epoch)})
-					return nil
-				}
-				return err
-			}
-		}
-		hash, err := b.challenge.Hash()
+		b.log.With().Info("building new challenge", log.EpochId(uint64(b.currentEpoch())))
+		err := b.buildNipstChallenge()
 		if err != nil {
-			return fmt.Errorf("getting challenge hash failed: %v", err)
+			return err
 		}
-		b.nipst, err = b.nipstBuilder.BuildNIPST(hash)
-		if err != nil {
-			return fmt.Errorf("cannot create Nipst: %v", err)
+	}
+	pubEpoch := b.challenge.PubLayerIdx.GetEpoch(b.layersPerEpoch)
+
+	hash, err := b.challenge.Hash()
+	if err != nil {
+		return fmt.Errorf("getting challenge hash failed: %v", err)
+	}
+	// ⏳ the following method waits for a PoET proof, which should take ~1 epoch
+	nipst, err := b.nipstBuilder.BuildNIPST(hash, b.layerClock.AwaitLayer((pubEpoch + 2).FirstLayer(b.layersPerEpoch)), b.stop)
+	if err != nil {
+		if _, stopRequested := err.(StopRequestedError); stopRequested {
+			return err
 		}
+		return fmt.Errorf("failed to build nipst: %v", err)
+	}
+
+	b.log.With().Info("awaiting publication epoch",
+		log.Uint64("pub_epoch", uint64(pubEpoch)),
+		log.Uint64("pub_epoch_first_layer", uint64(pubEpoch.FirstLayer(b.layersPerEpoch))),
+		log.Uint64("current_layer", uint64(b.layerClock.GetCurrentLayer())),
+	)
+	select {
+	case <-b.layerClock.AwaitLayer(pubEpoch.FirstLayer(b.layersPerEpoch)):
```

Comment:
have you tested what happens if the layer has passed already? will this not happen? also, why wait for the first layer of the publication epoch? shouldn't an ATX be pblished in `PubLayerIdx `

Reply:
What you're suggesting would change the existing behaviour and isn't justified at the moment. The only reason we wait here is to have a full view of the previous epoch and there's no reason for a miner to wait any longer. Barak even went further and said there's no incentive to even wait this long, but I implemented it this way because it improves performance (everyone having the same view) and there's not too much to gain from publishing sooner.

Actually, thinking about it a little, in the current implementation `PubLayerIdx` is always the first layer of the epoch, so it would actually be exactly the same.

Not sure what you mean by "not happen". `AwaitLayer` returns a closed (non-blocking) channel if it's called for a past layer, so if that's what you mean - we won't block if the layer passed. This feature of `AwaitLayer` is tested in the clock:
https://github.com/spacemeshos/go-spacemesh/blob/5198d32ae48cb92bc5c8c493b8e193a42a9e8893/timesync/clock_test.go#L181-L192